sdks/java/core/src/main/java/org/apache/beam/sdk/util/common/ReflectHelpers.java

```diff
@@ -223,12 +223,12 @@ public int compare(Object o1, Object o2) {
 
   /**
    * Finds the appropriate {@code ClassLoader} to be used by the
-   * {@link ServiceLoader#load} call, which by default would use the context
+   * {@link ServiceLoader#load} call, which by default would use the proposed
    * {@code ClassLoader}, which can be null. The fallback is as follows: context
    * ClassLoader, class ClassLoader and finaly the system ClassLoader.
    */
-  public static ClassLoader findClassLoader() {
-    ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
+  public static ClassLoader findClassLoader(final ClassLoader proposed) {
```

Comment:
Shouldn't the classloader ordering always be:
* Thread context classloader
* Current classloader
* System classloader

I could see that proposed would replace `ReflectHelpers.class.getClassLoader()`

Reply:
If beam would handle classloaders properly and without serialization yes, but since it doesnt and probably will not by desig  i think the transform owns the reference classloader first then tccl then system by delegation.

Reply:
I think its important that we stick with using the order I suggested because:
* people may run the DirectRunner for small jobs within an application container
* the SDK harness is becoming its own docker image and could support loading different DoFns with different classpaths.

Reply:
First point doesnt change much things since it will behave as a flat classpath main if you put beam in an app container.

Second point is my case but means this can be enhanced unwrapping the dofn/sdf from the transform with the logic of this PR. I can update it this way. Using the tccl would leak classes in the parent loader which is exactly the issue I'm trying to fix.

Reply:
Can you provide more details of how/when is the TCCL used in your usecase?

I can see that serializing objects becomes classloader aware and we wrap all serialized objects with environment specifications stating what was the classpath.

Reply:
Deploy bean in an app server - kind od tomcat-beam - and transforms/dofn in a webapp. Run pipelines, undeploy the app, redeploy it with another version of the fn, rerun the pipeline, it will fail and if luckily it doesnt then you leaked a proxy so a classloader so a lot of memory.

Reply:
Why is the TCCL not being cleared of classes when the webapp is undeployed if the TCCL is dedicated to the webapp and not shared across the webapp server?

If the TCCL is shared within the webapp server for common libraries then why are the DoFn or Apache Beam part of the shared resources?
Usually the TCCL is used for resources that are loadable and shared between multiple webapps like a SQL driver.

Reply:
Cause you have no guarantee at that point of the execution it is the webapp loader.

Side note: sql driver is unlikely a bad example, you can check all the effort tomcat does to not use the webapp loader since DriverManager can leak more or less the same way.

Reply:
Does "Cause you have no guarantee at that point of the execution it is the webapp loader." refer to my first question or second?

I'm not following your answer.

Reply:
Yes. In beam the tccl is not guaranteed at all and the webapp is just one part, the worker nodes are others and here agin tccl is undefined and undefinable. The DoFn classloader, however, is still valid and a secure location for the generated class.

Reply:
Also, the TCCL is controlled by the "user" since they control where it gets deployed/run. Different client/worker environments have different guarantees on what they supply as the TCCL and its best if we stay consistent in usage across all of them. `ReflectHelpers#findClassLoader` is used in several places and only changing one will lead to inconsistent behavior.

Reply:
No, it is really where the proxy is stored. If you use the tccl you still leak in most places for dofns/transforms and will finally OOME at some point if you reuse instances/machines. Only sane solution is to bind the proxy in the classloader the parent is. Also note it is the only way to proxy other methods than public ones due to java limitation.

Can do to handle dofn more deeply - next pr probably but this one puts proxies where they belong to in teem of classloader rules. Tccl is just random and not under beam or user control sadly in big data env so not reliable at all for a portable impl :(.